### PR TITLE
Bug 1871742: can't delete a PVC created using a data upload

### PR DESCRIPTION
**Pull request by glekner**
For Normal PVCs the modal is the same:
Screen Shot 2020-08-24 at 16 36 25

For CDI Bound PVCs the modal is:
Screen Shot 2020-08-24 at 16 36 32

@matthewcarleton how do we want this to look?

### CODE REVIEW

**Comment by glekner** [comment]
@suomiy @rawagner

**Comment by jelkosz** [comment]
I like it!

File: frontend/public/components/modals/delete-pvc-modal.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  );
+
+  return (
+    <form onSubmit={submit} className="modal-content">
+      <ModalTitle>
+        <YellowExclamationTriangleIcon className="co-icon-space-r" /> Delete Persistent Volume Claim
+      </ModalTitle>
+      <ModalBody>
+        <Stack hasGutter>
+          {alertComponents.map((alert, i) => (
```
Comment (rawagner) [inline frontend/public/components/modals/delete-pvc-modal.tsx:46]:
```suggestion
          {alertComponents}
```

File: frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/pvc-delete-extension.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import * as React from 'react';
+import { Alert, AlertVariant } from '@patternfly/react-core';
+import { k8sKill, K8sResourceKind } from '@console/internal/module/k8s';
+import { DataVolumeModel } from '../../models';
+
+export const killCDIBoundPVC = (pvc: K8sResourceKind) => k8sKill(DataVolumeModel, pvc);
+
+export const PVCDeleteAlertExtension: React.FC<PVCDeleteAlertExtension> = ({ pvc }) => (
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/pvc-delete-extension.tsx:8]: you should also consider that the extension would define `alertTitle: string`, `alertBody: LazyLoader<{pvc: PersistentVolumeClaimKind}>`, `alertType: AlertVariant`. 

It makes the extension less error prone, you dont have to specify className, isInline etc..all would be handled by pvc delete modal

**Review by rawagner — CHANGES_REQUESTED** [review]

**Comment by matthewcarleton** [comment]
> For Normal PVCs the modal is the same:
> Screen Shot 2020-08-24 at 16 36 25
> 
> For CDI Bound PVCs the modal is:
> Screen Shot 2020-08-24 at 16 36 32
> 
> @matthewcarleton how do we want this to look?

We need to ensure that if the PVC is used to back any OS then we highlight that here. It's possible that a user could delete this without knowing that's being cloned as an OS source somewhere else right?

File: frontend/public/components/modals/delete-pvc-modal.tsx (hunk 1 above)
Comment (glekner) [reply frontend/public/components/modals/delete-pvc-modal.tsx:46]: I want to render a margin between every alert `<StackItem>`, so I need this map.

**Comment by glekner** [comment]
fixed suggestions @rawagner 
@matthewcarleton well yes, deleting is a serious thing but I don't think we should check for every VM that uses the specific PVC. We could maybe check if that PVC is a golden one and let the user know its important?

**Comment by matthewcarleton** [comment]
> fixed suggestions @rawagner
> @matthewcarleton well yes, deleting is a serious thing but I don't think we should check for every VM that uses the specific PVC. We could maybe check if that PVC is a golden one and let the user know its important?

Ya that's what I was thinking. If it's that important we shouldn't assume the user is thinking about it that way. It would be a nice addition to alert them that is being used to support and OS as a backing source.

**Comment by glekner** [comment]
@matthewcarleton can you approve the text here?
Screen Shot 2020-08-30 at 12 27 13

**Comment by jelkosz** [comment]
@glekner I dont think it will cause any malfunction. We clone the image. It will cause that the other VMs will not be able to use it but the ones which have already been deployed will be fine. Or am I missing something?

File: frontend/public/components/modals/delete-pvc-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    const extensionPromises = pvcDeleteExtensions.map(
+      ({ properties: { predicate, onPVCKill } }) => predicate(pvc) && onPVCKill(pvc),
+    );
+    return handlePromise(Promise.all([promise, ...extensionPromises]), close);
+  };
+
+  const alertComponents = pvcDeleteExtensions.map(
+    ({ properties: { predicate, alert }, uid }) =>
+      predicate(pvc) && (
+        <Alert
```
Comment (rawagner) [inline frontend/public/components/modals/delete-pvc-modal.tsx:34]: If you wrap the alert in `StackItem` you wont need to `map` again later.

```suggestion
       <StackItem key={uid}>
          <Alert
```
you can also move key from `Alert` to `StackItem`. There's also no need to make `key` more complicated than just uid. Doing `pvc-alert-${uid}` doesnt bring any value

File: frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/pvc-delete-extension.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+export const killCDIBoundPVC = (pvc: PersistentVolumeClaimKind) =>
+  k8sKill(DataVolumeModel, {
+    metadata: {
+      name: pvc?.metadata?.name,
+      namespace: pvc?.metadata?.namespace,
+    },
+  });
+
+export const PVCDeleteAlertExtension = ({ pvc }) => {
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/pvc-delete-extension.tsx:19]:
```suggestion
export const PVCDeleteAlertExtension: React.FC<{ pvc: PersistentVolumeClaimKind }> = ({ pvc }) => {
```

File: frontend/packages/console-plugin-sdk/src/typings/pvc.ts [hunk 5]
```diff
@@ -22,6 +23,22 @@ namespace ExtensionProperties {
...
+
+  export interface PVCDelete {
+    /** predicate that tells if to use the extension or not */
+    predicate: (pvc: PersistentVolumeClaimKind) => boolean;
+    /** method for the pvc delete operation */
+    onPVCKill: (pvc: PersistentVolumeClaimKind) => Promise<void>;
+    /** alert for additional info */
+    alert?: {
+      /** alert title */
+      alertTitle: string;
```
Comment (rawagner) [inline frontend/packages/console-plugin-sdk/src/typings/pvc.ts:35]: could be just `title`, 'type', 'Body'. Not a blocker though :)

File: frontend/public/components/modals/delete-pvc-modal.tsx (hunk 1 above)
Comment (rawagner) [reply frontend/public/components/modals/delete-pvc-modal.tsx:53]: See my comment https://github.com/openshift/console/pull/6407#discussion_r480074555

**Comment by matthewcarleton** [comment]
> @matthewcarleton can you approve the text here?
> Screen Shot 2020-08-30 at 12 27 13

Much better!
The only thing I would ask is if we can provide which OS it's backing. It will help the user make the connection easier.

**Comment by rawagner** [comment]
/hold

needs rebase

**Pull request merged**